### PR TITLE
Users/prachisingla/onboarding CTS traffic WL 

### PR DESCRIPTION
**Pull request by psingla1210**

### CODE REVIEW

File: website/docs/workloads/ctstraffic/ctsTrafficUsage.docx
Comment (yangpanMS) [inline website/docs/workloads/ctstraffic/ctsTrafficUsage.docx:1]: Is there a reason this is doc? Could we do markdown?

Comment (yangpanMS) [inline website/docs/workloads/ctstraffic/ctsTrafficUsage.docx:1]: I see therere is a url for the docx file. For things that's not us, you could just leave a link in markdown, we don't need to put docx in source

File: website/docs/workloads/ctstraffic/ctstraffic.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# CtsTraffic
+ctsTraffic is a highly scalable client/server networking tool giving detailed performance and reliability analytics
+
+* [Official CtsTraffic Documentation](https://github.com/microsoft/ctsTraffic/tree/master/documents)
+* [Official Release](https://github.com/microsoft/ctsTraffic/tree/master/Releases/2.0.3.0)
+
+## What is Being Measured?
+ctsTraffic is designed as a classic client/server model, where the server will wait listening for client requests, while the client initiates connections to servers. ctsTraffic servers will accept any number of connections from any number of clients; ctsTraffic clients can target and make connections to one or more servers.
```
Comment (yangpanMS) [inline website/docs/workloads/ctstraffic/ctstraffic.md:8]: how is this different from the ntttcp tests? for bandwidth as example, which one should we use?

File: src/VirtualClient/VirtualClient.Packaging/win-x64/ctstraffic/ctsTraffic.exe
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Packaging/win-x64/ctstraffic/ctsTraffic.exe:1]: We don't need to put it in source. We can just keep it in storage account.

File: src/VirtualClient/build-packages-workloads.cmd [hunk 2]
```diff
@@ -23,6 +23,11 @@ if /i "%PackageName%" NEQ "" (
     call dotnet pack %PackagingProject% --force --no-restore --no-build -c Debug -p:NuspecFile=%PackagingDir%\nuspec\%PackageName%.nuspec && echo: || Goto :Error
 
 ) else (
+    echo:
```
Comment (yangpanMS) [inline src/VirtualClient/build-packages-workloads.cmd:26]: We don't need to package nupkg for this

File: src/VirtualClient/VirtualClient.Packaging/nuspec/VirtualClient.nuspec [hunk 3]
```diff
@@ -27,6 +27,8 @@
         =============================================================================================
         -->
         <file src="$BinConfigDir$\x64\VirtualClient.Main\net6.0\win-x64\publish\**\*" target="content\win-x64" />
+		<file src="$ToolsOutDir$\win-x64\ctstraffic\**\*" target="content\win-x64\packages\ctstraffic\win-x64" />
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Packaging/nuspec/VirtualClient.nuspec:30]: Is there a reason we need this in vc package directly?

File: src/VirtualClient/VirtualClient.Main/VirtualClient.Main.csproj [hunk 4]
```diff
@@ -77,6 +77,7 @@
         <WindowsProfiles Include="profiles\PERF-IO-FIO-MULTITHROUGHPUT.json" />
         <WindowsProfiles Include="profiles\PERF-NETWORK-PING.json" />
         <WindowsProfiles Include="profiles\PERF-NETWORK.json" />
+		<WindowsProfiles Include="profiles\PERF-NETWORK-CTSTRAFFIC.json" />
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Main/VirtualClient.Main.csproj:80]: nit: spacing

File: src/VirtualClient/VirtualClient.Actions/CtsTraffic/CtsTrafficMetricsParser.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+                            {
+                                // do nothing as this result file has non-double values.
+                            }
+
+                            try
+                            {
+                                // cumulative count of failed IO patterns due to Winsock errors
+                                this.Metrics.Add(new Metric($"NetworkError(TimeSlice-{values[0].Trim()})", Convert.ToDouble(values[5]), MetricRelativity.LowerIsBetter));
+                            }
+                            catch
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/CtsTraffic/CtsTrafficMetricsParser.cs:95]: why are there so much catch blocks? Is the output pretty flaky?

File: src/VirtualClient/VirtualClient.Actions.UnitTests/CtsTraffic/CtsTrafficMetricsParserTests.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            MetricAssert.Exists(metrics, "InFlight(TimeSlice-0.006)", 1);
+            MetricAssert.Exists(metrics, "Completed(TimeSlice-0.006)", 0);
+            MetricAssert.Exists(metrics, "NetworkError(TimeSlice-0.006)",0);
+            MetricAssert.Exists(metrics, "DataError(TimeSlice-0.006)", 0);
+            MetricAssert.Exists(metrics, "SendBps(TimeSlice-20.007)", 27375465, "B/s");
+            MetricAssert.Exists(metrics, "RecvBps(TimeSlice-20.007)", 119906898, "B/s");
+            MetricAssert.Exists(metrics, "InFlight(TimeSlice-20.007)", 1);
+            MetricAssert.Exists(metrics, "Completed(TimeSlice-20.007)", 0);
+            MetricAssert.Exists(metrics, "NetworkError(TimeSlice-20.007)", 0);
+            MetricAssert.Exists(metrics, "DataError(TimeSlice-20.007)", 0);
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions.UnitTests/CtsTraffic/CtsTrafficMetricsParserTests.cs:51]: Would you please help add at least the first and last rows if we are not verifying full 30

**Review by yangpanMS — APPROVED** [review]
The whole PR looks good, but please help me understand when will this becomes a better workload than ntttcp and sockperf/latte

File: website/docs/workloads/ctstraffic/ctstraffic.md (hunk 1 above)
Comment (psingla1210) [reply website/docs/workloads/ctstraffic/ctstraffic.md:8]: Target of NTTTCP/Latte and ctstraffic is different. Former ones are more inclined towards performance where as ctstraffic is more for reliability. It is a simple benchmark which is highly scalable with a lot of options to control time, number of clients, number of servers, exact buffer and total data etc. that can be run in earlier phases of HW qualification/scenarios where we are more concerned about reliability of networking among several machines, we dont care how they performed, we just care that they should be able to communicate and able to send and receive data.

File: src/VirtualClient/VirtualClient.Actions.UnitTests/CtsTraffic/CtsTrafficMetricsParserTests.cs (hunk 6 above)
Comment (psingla1210) [reply src/VirtualClient/VirtualClient.Actions.UnitTests/CtsTraffic/CtsTrafficMetricsParserTests.cs:51]: I have already done the same.

File: src/VirtualClient/VirtualClient.Actions/CtsTraffic/CtsTrafficMetricsParser.cs (hunk 5 above)
Comment (psingla1210) [reply src/VirtualClient/VirtualClient.Actions/CtsTraffic/CtsTrafficMetricsParser.cs:95]: Yes, there is high possibility that one of the column is giving non-double value as this workload works like that, still i want to be sure that flow should not go in catch and further columns should be recorded in the telemetery.

File: website/docs/workloads/ctstraffic/ctsTrafficUsage.docx
Comment (psingla1210) [reply website/docs/workloads/ctstraffic/ctsTrafficUsage.docx:1]: True that! Deleted.

File: src/VirtualClient/VirtualClient.Packaging/nuspec/VirtualClient.nuspec (hunk 3 above)
Comment (psingla1210) [reply src/VirtualClient/VirtualClient.Packaging/nuspec/VirtualClient.nuspec]: nope, removed.

File: src/VirtualClient/build-packages-workloads.cmd (hunk 2 above)
Comment (psingla1210) [reply src/VirtualClient/build-packages-workloads.cmd]: updated

File: src/VirtualClient/VirtualClient.Packaging/ctstraffic.vcpkg [hunk 7]
```diff
@@ -0,0 +1,6 @@
+﻿{
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Packaging/ctstraffic.vcpkg:1]: You have weird char here

File: src/VirtualClient/VirtualClient.Actions/CtsTraffic/CtsTrafficMetricsParser.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                            }
+
+                            try
+                            {
+                                // cumulative count of successfully completed IO patterns
+                                this.Metrics.Add(new Metric($"Completed(TimeSlice-{values[0].Trim()})", Convert.ToDouble(values[4]), MetricRelativity.HigherIsBetter));
+                            }
+                            catch
+                            {
+                                // do nothing as this result file has non-double values.
```
Comment (yangpanMS) [inline src/VirtualClient/VirtualClient.Actions/CtsTraffic/CtsTrafficMetricsParser.cs:87]: If you may, could you wrap these try-metricsAdd-catch blocks into a method? Just to make this method less bulky

**Review by yangpanMS — APPROVED** [review]

File: src/VirtualClient/VirtualClient.Packaging/ctstraffic.vcpkg (hunk 7 above)
Comment (psingla1210) [reply src/VirtualClient/VirtualClient.Packaging/ctstraffic.vcpkg:1]: Can u still see it? I dont.

Comment (yangpanMS) [reply src/VirtualClient/VirtualClient.Packaging/ctstraffic.vcpkg:1]: image

Comment (psingla1210) [reply src/VirtualClient/VirtualClient.Packaging/ctstraffic.vcpkg]: It is gone now

**Review by yangpanMS — APPROVED** [review]

**Review by yangpanMS — APPROVED** [review]

**Pull request merged**